ingestion-core/src/main/java/com/mozilla/telemetry/ingestion/core/transform/PubsubMessageToObjectNode.java

```diff
@@ -595,6 +642,59 @@ private String getAndCacheBqName(String name) {
       }
     }
 
+    /**
+     * Encode the given JSON object as a compact string.
+     *
+     * <p>We maintain a BigQuery persistent UDF that can parse a variety of formats, so any format
+     * used here must first be supported in:
+     * https://github.com/mozilla/bigquery-etl/blob/master/mozfun/hist/extract/udf.sql
+     *
+     * <p>Schema validation should generally ensure that all histograms are well-formed, but we
+     * perform some light sanity checking on the values and fall back to encoding the histogram
+     * as a JSON string if anything seems off.
+     */
+    private TextNode compactHistogramEncoding(JsonNode o, String fieldName) {
+      TextNode jsonEncoding = TextNode.valueOf(Json.asString(o));
+      int histogramType = o.path("histogram_type").asInt(-1);
+      if (histogramType < 0 || histogramType > 5) {
+        incrementInvalidHistogramType();
+        return jsonEncoding;
+      }
+      long sum = o.path("sum").asLong(-1);
+      if (sum < 0) {
+        incrementInvalidHistogramSum();
+        return jsonEncoding;
+      }
+      JsonNode values = o.path("values");
+      if (histogramType == 4) {
+        // Type 4 histograms are "count" histograms representing only a single value.
+        return TextNode.valueOf(Long.toString(sum));
+      } else if (histogramType == 2 && fieldName.startsWith("use_counter2")) {
+        // "use counters" represent the bulk of histogram data sent in main pings, so we provide
+        // additional optimization for this case; although they are encoded as "boolean" histograms
+        // (type 2), the "0" (false) bucket is never populated, meaning they only encode a single
+        // value.
+        return TextNode.valueOf(Long.toString(sum));
```

Comment:
I'd rather not have a duplicate `TextNode.valueOf(Long.toString(sum))` here:
```suggestion
      if (histogramType == 4 || (histogramType == 2 && fieldName.startsWith("use_counter2"))) {
        // Type 4 histograms are "count" histograms representing only a single value.
        // "use counters" represent the bulk of histogram data sent in main pings, so we provide
        // additional optimization for this case; although they are encoded as "boolean" histograms
        // (type 2), the "0" (false) bucket is never populated, meaning they only encode a single
        // value.
        return TextNode.valueOf(Long.toString(sum));
```